fence/resources/openid/ras_oauth2.py

```
@@ -0,0 +1,75 @@
+            idp="ras",
+            HTTP_PROXY=HTTP_PROXY,
+        )
+
+    def get_auth_url(self):
+        """
+        Get authorization uri from discovery doc
+        """
+        authorization_endpoint = self.get_value_from_discovery_doc(
+            "authorization_endpoint", ""
+        )
+
+        uri, state = self.session.create_authorization_url(
+            authorization_endpoint, prompt="login"
+        )
+
+        return uri
+
+    def get_userinfo(self, token, userinfo_endpoint, code):
+        access_token = token["access_token"]
+        header = {"Authorization": "Bearer " + access_token}
+        res = requests.get(userinfo_endpoint, headers=header)
+        return res.json()
+
+    def get_user_id(self, code):
+        try:
+            token_endpoint = self.get_value_from_discovery_doc("token_endpoint", "")
+            jwks_endpoint = self.get_value_from_discovery_doc("jwks_uri", "")
+            userinfo_endpoint = self.get_value_from_discovery_doc(
+                "userinfo_endpoint", ""
+            )
+
+            token = self.get_token(token_endpoint, code)
+            keys = self.get_jwt_keys(jwks_endpoint)
+
+            claims = jwt.decode(
+                token["id_token"],
+                keys,
+                options={"verify_aud": False, "verify_at_hash": False},
+            )
+
+            userinfo = self.get_userinfo(token, userinfo_endpoint, code)
+
+            if "preferred_username" in userinfo and userinfo["preferred_username"]:
+                return {"ras": userinfo["preferred_username"]}
+            elif "UserID" in userinfo and userinfo["UserID"]:
+                return {"ras": userinfo["UserID"]}
+            elif claims["sub"]:
+                return {"ras": claims["sub"]}
+            else:
+                return {"error": "Can't get user's ras"}
+        except Exception as e:
+            self.logger.exception("Can't get user info")
+            return {"error": "Can't get your ras: {}".format(e)}
```

a couple comments here
- we don't want to return the full error to the user (`"Can't get your ras: {}".format(e)`)
- it's good practice to limit the number of return statements; i think this would be easier to read/debug
```
err_msg = "Can't get user's info"
try:
    ...

    username = None
    if userinfo.get("preferred_username"):
        username = userinfo["preferred_username"]
    elif userinfo.get("UserID"):
        username = userinfo["UserID"]
    elif claims.get("sub"):
        username = claims["sub"]
    if not username:
        logger.error("{}, received claims: {} and userinfo: {}".format(err_msg, claims, userinfo))
        return {"error": err_msg}
    return {"username": username}
except Exception as e:
    self.logger.exception("{}: {}".format(err_msg, e))
    return {"error": err_msg}
```